File: task9/task9.cpp
#include <iostream>
#include <vector>
#include <stack>

std::string brainLuck(const std::string &code, const std::string &input)
{
    int len_input = input.length();
    int len_code = code.length();

    if (len_input == 0)
    {
        std::cout << "Input was empty!\n";
        return "";
    }

    if (len_code == 0)
    {
        std::cout << "Code was empty!\n";
        return input;
    }

    // col1: indices of openning brackets
    // col2: indices of closing brackets
    // for ,[.[-],] this is [[1, 7], [3, 5]]
    std::vector<std::vector<int>> brackets;

    std::string result_string;
    // result_string.resize(len_input, '\0');

    std::string::const_iterator it_input = input.cbegin();

    for (int i = 0; i < len_code; ++i)
    {
        if (code[i] == '[')
        {
            brackets.push_back({i, -1});
        }
        else if (code[i] == ']')
        {
            for (int j = brackets.size() - 1; j >= 0; --j)
            {
                if (brackets[j][1] == -1)
                {
                    brackets[j][1] = i;
                    break;
                }
            }
        }
    }

    char data_pointer = '0';
    for (int i = 0; i < len_code; ++i)
    {
        switch (code[i])
        {
        case ',':
        {
            data_pointer = *it_input;
            ++it_input;

            break;
        }
        case '.':
        {
            result_string += data_pointer;

            break;
        }
        case '+':
        {
            data_pointer = (data_pointer + 1) % 256;

            break;
        }
        case '-':
        {
            data_pointer = (char)(unsigned((int)data_pointer - 1)) % 256;

            break;
        }
        case '[':
        {
            if ((int)data_pointer == 0)
            {
                for (auto b : brackets) {
                    if (b[0] == i) {
                        i = b[1];
                        break;
                    }
                }
            }

            break;
        }
        case ']':
        {
            if ((int)data_pointer != 0)
            {
                for (auto b : brackets) {
                    if (b[1] == i) {
                        i = b[0];
                        break;
                    }
                }
            }

            break;
        }
        default:
        {
            std::cout << "ERROR: Unsupported symbol: " << code[i] << " (" << int(code[i]) << ')' << '\n';
            return "";
        }
        }
    }

    return result_string;
}

int main()
{
    //echo until "255";
    std::string tw = "c";
    tw.append(1, (char)255);
    std::cout << brainLuck(",.", tw) << '\n'; // "c"

    //echo until "255";
    tw = "c";
    tw.append(1, (char)255);
    std::cout << brainLuck(",-.", tw) << '\n'; // "c"

    //echo until "255";
    tw = "c";
    tw.append(1, (char)255);
    std::cout << brainLuck(",+.", tw) << '\n'; // "d"

    //echo until "255";
    tw = "c0";
    tw.append(1, (char)255);
    std::cout << brainLuck(",.,[.", tw) << '\n'; // "cc"

    // //echo until "255";
    // tw = "codewars";
    // tw.append(1, (char)255);
    // std::cout << brainLuck(",+[-.,+]", tw) << '\n'; // "codewars"

    //echo until "0";
    // std::string mw = "codewars";
    // mw.append(1, (char)0);
    // // endless recursion
    // std::cout << brainLuck(",[.[-],]", mw) << '\n'; // "codewars"

    //two number multiplier
    // std::string dw;
    // dw.append(1, (char) 7);
    // dw.append(1, (char) 3);
    // std::string result_string;
    // result_string.append(1, (char)21);
    // std::cout << brainLuck(",>,<[>[->+>+<<]>>[-<<+>>]<<<-]>>.",dw << '\n'; //

    return 0;
}

File: task9/task9.py
def bracket_indices(code, code_len):
    brackets_op = [
        idx
        for idx, bracket in enumerate(code)
        if bracket == '['
    ]

    brackets_cl = [
        code_len - idx - 1
        for idx, bracket in enumerate(code[::-1])
        if bracket == ']'
    ]

    indices = dict(zip(brackets_op, brackets_cl))
    return indices

def brain_luck(code, program_input):
    code_len = len(code)
    code_idx = 0
    
    input_len = len(program_input)
    input_idx = 0

    brackets = bracket_indices(code, code_len)
    brackets_inverted = {value: key for key, value in brackets.items()}

    current_char = '0'

    result = []

    while code_idx < code_len:
        if code[code_idx] == '>':
            pass
        elif code[code_idx] == '<':
            pass
        elif code[code_idx] == '+':
            current_char = chr((ord(current_char) + 1) % 256)
        elif code[code_idx] == '-':
            current_char = chr((ord(current_char) - 1) % 256)
        elif code[code_idx] == '.':
            result.append(current_char)
        elif code[code_idx] == ',':
            current_char = program_input[input_idx]
            input_idx += 1
        elif code[code_idx] == '[':
            if ord(current_char) == 0:
                code_idx = bracket_indices.get(code_idx, None)
        elif code[code_idx] == ']':
            if ord(current_char) != 0:
                code_idx = brackets_inverted.get(code_idx, None)
        else:
            print(f'Unsupported instruction: "{code[code_idx]}"!')
            return ''
        
        code_idx += 1

    return ''.join(result)


def main():
    # Echo until byte(255) encountered
    print(brain_luck(",+[-.,+]", "Codewars" + chr(255)))  # "Codewars"

    # Echo until byte(0) encountered
    print(brain_luck(",[.[-],]", "Codewars" + chr(0)))  # "Codewars"

    # Two numbers multiplier
    print(brain_luck(",>,<[>[->+>+<<]>>[-<<+>>]<<<-]>>.", chr(8) + chr(9)))  # chr(72)


if __name__ == "__main__":
    main()
